src/FreeformIdeas.js

```diff
@@ -1,90 +1,173 @@
+  const [linkingState, setLinkingState] = useState(false);
+  const linkBreak = useCallback(() => setLinkingState(false), []);
+
+  const onLinkStart = useCallback((e) => {
+    setLinkerEnd([e.clientY - 65, e.clientX]);
+    setLinkingState(true);
+  }, []);
+
+  const adjustLinker = useCallback(
+    (e) => {
+      if (linkingState) {
+        setLinkerEnd([e.clientY - 65, e.clientX]);
+      }
+    },
+    [linkingState]
+  );
+
+  const onLinkEnd = useCallback(
+    (id) => {
+      if (selectedIdea.id !== id && linkingState) {
+        if (!selectedIdea.linkList.includes(id)) {
+          selectedIdea.linkList.push(id);
+        }
+      }
+      setLinkingState(false);
+    },
+    [selectedIdea, linkingState]
+  );
 
   const [, drop] = useDrop({
     accept: ItemTypes.IDEA,
     drop(item, monitor) {
       const delta = monitor.getDifferenceFromInitialOffset();
       const left = Math.round(item.left + delta.x);
       const top = Math.round(item.top + delta.y);
-      setIdeas(
-        update(ideas, {
-          [item.id]: {
-            $merge: { left, top },
-          },
-        })
-      );
+      ideasDispatch({ type: "update", id: item.id, data: { left, top } });
     },
   });
 
   const editIdea = useCallback(
     (id) => {
-      setSelectedIdea(ideas[id]);
+      setSelectedId(id);
       ideaModalShow(true);
     },
-    [ideas, ideaModalShow]
+    [setSelectedId, ideaModalShow]
   );
 
   const handleDoubleClick = useCallback(
     (e) => {
       if (e.target.className !== "FreeformMap") return;
       e.preventDefault();
-      setSelectedIdea({
+      setCoords({
         top: e.nativeEvent.layerY,
         left: e.nativeEvent.layerX,
-        title: "",
       });
+      setSelectedId(uuidv4());
       setTimeout(ideaModalShow, 300);
     },
     [ideaModalShow]
   );
 
   const handleIdeaChange = useCallback(
     (idea) => {
+      if (idea.title === "") return;
+      setShowIdeaModal(false);
       if (!idea.id) {
-        idea = update(idea, { id: { $set: uuidv4() } });
-        setSelectedIdea(idea);
+        ideasDispatch({
+          type: "create",
+          id: selectedId,
+          data: update(idea, { $merge: coords }),
+        });
+        return;
       }
-      const { id, title } = idea;
-      if (title === "") return;
-      setIdeas(
-        update(ideas, {
-          [id]: {
-            $set: idea,
-          },
-        })
-      );
-      setShowIdeaModal(false);
+      ideasDispatch({ type: "update", id: selectedId, data: idea });
     },
-    [ideas]
+    [coords, selectedId]
   );
 
+  function calcCoords(height, width, arrowRotation, ideaTop, ideaLeft) {
+    let top = 0;
+    let left = 0;
+    const referenceAngleOne =
+      (180 / Math.PI) * Math.atan2(height / 2, width / 2);
+    const referenceAngleTwo = 180 - referenceAngleOne;
+    if (
+      Math.abs(arrowRotation) > referenceAngleOne &&
+      Math.abs(arrowRotation) < referenceAngleTwo
```

Comment:
Save `Math.abs(arrowRotation)` to a variable instead of recalculating it 4 times.